### ISSUE TITLE
Support a common dependencies feature

### ISSUE THREAD

**Issue by arackaf** [systemjs/builder#528]
Consider:

module1
   Uses: ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h']
module2
   Uses: ['b', 'c', 'g', 'h', 'k']
module3
   Uses: ['g', 'h', 'i', 'k']

It would potentially be useful to create a bundle of all duplicated dependencies across any of module1, module2, and module3, which from the above would be ['b', 'c', 'g', 'h', 'k'].  Call that bundle X.

Modules 1, 2, and 3 could then subtract X from their builds.  Of course this means the bundles would wind up loading up some things at runtime in X that they don't need (module1 would load 'k', module3 would load 'b', 'c', and probably others).

Would this be a useful feature?  Would you be interested in a PR implementing it?

**Comment by arackaf** [comment]
Looking through your code, and thinking about it, *if* this were to be supported as an arithmetic expression, and not just as a method on builder, getting the syntax right would be tricky.  This operation would have to be greedy, ie (assuming % represented this operation)

A % B % C

would have to be executed up front, on all three.  You'd have to get the trees for A B and C, and analyze them all at once, keeping anything duplicated between any of them.  I think the only real way to do that would be to first implement parens, which you already have an issue for, and which I'd been looking through your code trying to see if I could knock out, and then support it like this

(% A B C)

so basically add LISP syntax, which you may or may not be happy with.  And of course you could combine with other expressions

X - (% A B C)

Or of course it could just be a method on builder.

**Comment by arackaf** [comment]
Now that parens are supported, I'm hoping there's a particular syntax for this you'd be amenable to.  If so I'd look to start implementing.

**Comment by guybedford** [comment]
I do think this is actually more of a bundle strategy, which may not be suited to the current binary operator syntax.

Rather such a feature ("all modules in A, B and C that belong to more than one of each tree") would go in the same category sort of feature as "all modules in A, B and C that are larger than 1KB" or "all modules in A, B and C under the app/ path".

These kinds of strategies may be best implemented using custom tracing methods:

```javascript
var treeA = builder.trace('A');
var treeB = builder.trace('B');
var treeC = builder.trace('C');

builder.bundle(customStrategy(treeA, treeB, treeC));
```

**Comment by guybedford** [comment]
The only way I can think of to formalize this would be by creating a general syntax for functions:

```
builder.bundle('$common(A, B, C) + D + E', {
  strategies: {
    common(trees) {
      var commonModules = {};
      trees.forEach((tree, index) => {
        var treeModules = Object.keys(tree);
        treeModules.forEach((module) => {
          var load = tree[module];
          if (typeof load != 'object')
            return;
          var inAnother = trees.some((tree, curIndex) => {
              if (index == curIndex)
                return false;
              return Object.keys(tree).indexOf(module) != -1;
            });
          });
          if (inAnother) {
            commonModules[module] = load;
          }
        });
      });
      return commonModules;
    }
  }
})
```

The strategy itself could then be turned into a module, and used with:

```javascript
builder.bundle('$common(A, B, C) + D + E', {
  strategies: {
    common: require('systemjs-builder-strategy-common')
  }
});
```

**Comment by guybedford** [comment]
But certainly territory to tread carefully in wrt implementation!

**Comment by arackaf** [comment]
That would be ideal, actually.  Very nice idea.  Are these "strategies" documented anywhere???

**Comment by arackaf** [comment]
And of course I'd add some tests to make sure

```javascript
builder.bundle('$file1 + $file2')
```

still worked appropriately.

**Comment by arackaf** [comment]
Also - I assume strategies referenced in arithmetic syntax would NOT be canonicalized or expanded, right?

`'$common('a + b + c')`

would be given literally `'a + b + c'` to handle however it wants, right?

**Comment by guybedford** [comment]
I'm by no means advocating implementation here, simply playing around with ideas. There are massive ecosystem consequences to be considered first.

In the suggestion functions take comma separated parameters, which are themselves expressions.

**Comment by arackaf** [comment]
For sure - don't worry, I won't rush into anything.

That said, can you elaborate on this "strategy" feature - I don't recall (and still don't having just looked) seeing it documented anywhere.  It seems that either way I could write my own strategy as a starting point, even if it's not initially (or ever) able to be referenced in an expression.

**Comment by arackaf** [comment]
So it looks like you were throwing this out as a potential new feature - ie, let the user specify strategies that are then accessible within builder?

Why not make it even simpler

`builder.bundle('$common(A, B, C) + D + E')`

which then just calls `builder.common` with arguments `'A, B, C'`.  And of course the user would be responsible for adding a `common` method onto the builder's prototype, which could be done in any number of ways, including potentially something so simple as `require('builder-common-addon')` which would do the busy work for us.

Thought?

**Comment by guybedford** [comment]
@arackaf yes exactly. Plugin patterns that mutate modules aren't actually good idea in true modular environments. While collissions are very rare, it's still creating a global namespace, which should be avoided where possible.
